packages/monitor-v2/src/monitor-og/oSnapAutomation.ts

```diff
@@ -0,0 +1,393 @@
+        const rulesVerified = verifyRules(
+          getModuleParameters(proposal.plugins.safeSnap.safes[0].umaAddress, supportedModules).parsedRules,
+          proposal
+        ).verified;
+
+        // Return verification result together with original proposal. This is used for filtering below.
+        return { verified: voteOutcomVerified && ipfsVerified && rulesVerified, proposal };
+      })
+    )
+  ).filter((proposal) => proposal.verified); // Filter out all proposals that did not pass verification.
+
+  // Convert back to SnapshotProposalExpanded format.
+  return verifiedProposals.map((verificationResult) => {
+    const proposal = verificationResult.proposal;
+    const { plugins, ...clonedObject } = proposal;
+    return { ...clonedObject, safe: proposal.plugins.safeSnap.safes[0] };
+  });
+};
+
+const approveBond = async (
+  provider: Provider,
+  signer: Signer,
+  currency: string,
+  bond: string,
+  spender: string
+): Promise<void> => {
+  // If bond is 0, no need to approve.
+  if (bond === "0") return;
+
+  // If existing approval matches the bond, no need to proceed.
+  const currencyContract = await getContractInstanceWithProvider<ERC20Ethers>("ERC20", provider, currency);
+  const currentAllowance = await currencyContract.allowance(await signer.getAddress(), spender);
+  if (currentAllowance.toString() === bond) return;
+
+  try {
+    await (await currencyContract.connect(signer).approve(spender, bond)).wait();
+  } catch (error) {
+    assert(error instanceof Error, "Unexpected Error type!");
+    throw new Error(`Bond approval for ${spender} failed: ${error.message}`);
+  }
+};
+
+const submitProposals = async (
+  logger: typeof Logger,
+  proposals: SnapshotProposalExpanded[],
+  supportedModules: SupportedModules,
+  params: MonitoringParams
+) => {
+  assert(params.signer !== undefined, "Signer must be set to propose transactions.");
+
+  for (const proposal of proposals) {
+    const og = await getOgByAddress(params, proposal.safe.umaAddress);
+
+    // Approve bond based on stored module parameters.
+    const moduleParameters = getModuleParameters(proposal.safe.umaAddress, supportedModules);
+    await approveBond(params.provider, params.signer, moduleParameters.currency, moduleParameters.bond, og.address);
+
+    // Create transaction parameters.
+    const transactions = proposal.safe.txs.map((transaction) => {
+      return {
+        to: transaction.mainTransaction.to,
+        operation: transaction.mainTransaction.operation,
+        value: transaction.mainTransaction.value,
+        data: transaction.mainTransaction.data,
+      };
+    });
+    const explanation = ethersUtils.toUtf8Bytes(proposal.ipfs);
+
+    // Check that proposal submission would succeed.
+    try {
+      await og.callStatic.proposeTransactions(transactions, explanation, { from: await params.signer.getAddress() });
+    } catch (error) {
+      assert(error instanceof Error, "Unexpected Error type!");
```

Comment:
What kind of errors could we have that are not instance of Error? Have you experienced any?

Reply:
Assert here is more to allow type safe accessing of error.message later down when composing wrapped error.